keps/sig-storage/NNNN-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1017 @@
+   - `OnDelete`: the default value, only update the PVC when the the old PVC is deleted.
+   - `InPlace`: update the PVC in-place if possible. Also includes the `OnDelete` behavior.
+
+4. Introduce a new field in StatefulSet `spec.updateStrategy.rollingUpdate`: `volumeClaimSyncStrategy`
+   to specify how to update PVCs and Pods. Possible values are:
+   - `Async`: the default value, preseve the current behavior.
+   - `LockStep`: update PVCs first, then update Pods. See below for details.
+
+5. Collect the status of managed PVCs, and show them in the StatefulSet status.
+
+### Updated Reconciliation Logic
+
+How to update PVCs:
+1. If `volumeClaimUpdateStrategy` is `InPlace`,
+   and if `volumeClaimTemplate` and actual PVC only differ in mutable fields
+   (`spec.resources.requests.storage`, `spec.volumeAttributesClassName`, `metadata.labels`, and `metadata.annotations` currently),
+   update the PVC in-place to the extent possible.
+   Do not perform the update that will be rejected by API server, such as
+   decreasing the storage size below its current status.
+   Note that decrease the size can help recover from a failed expansion if 
+   `RecoverVolumeExpansionFailure` feature gate is enabled.
+
+2. If it is not possible to make the PVC [compatible](#what-pvc-is-compatible),
+   do nothing. But when recreating a Pod and the corresponding PVC is deleting,
+   wait for the deletion then create a new PVC with the current template
+   together with the new Pod (already implemented).
+<!--
+Tested on Kubernetes v1.28, and I can see this event:
+Warning  FailedCreate         3m58s (x7 over 3m58s)  statefulset-controller  create Pod test-rwop-0 in StatefulSet test-rwop failed error: pvc data-test-rwop-0 is being deleted
+-->
+
+When to update PVCs:
+1. Before advancing `status.updatedReplicas` to the next replica,
+   additionally check that the PVCs of the next replica are 
+   [compatible](#what-pvc-is-compatible) with the new `volumeClaimTemplate`.
+   If not, update the PVC after old Pod deleted, before creating new pod,
+   or if update is not possible:
+   <!-- TODO: what if the update to PVC failed? -->
+   - If `volumeClaimSyncStrategy` is `LockStep`,
+     wait for the user to delete/update the old PVC manually.
+   - If `volumeClaimSyncStrategy` is `Async`,
+     the diff is ignored and the normal rolling update proceeds.
+
+2. If Pod spec does not change, only mutable fields in `volumeClaimTemplate` differ,
+   The PVCs should be updated just like Pods would. A replica is considered ready
+   if all its volumes are compatible with the new `volumeClaimTemplate`.
+   `spec.ordinals` and `spec.updateStrategy.rollingUpdate.partition` are also respected.
+   e.g.:
+   - If `spec.updateStrategy.type` is `RollingUpdate`,
+     update the PVCs in the order from the largest ordinal to the smallest.
+     Only proceed to the next ordinal when all the PVCs of the previous ordinal
```

Comment:
By saying compatible, we are already waiting for the resize to finish. Because in "What PVC is compatible", `status.capacity.storage` should be updated before a PVC becomes compatible.